packages/client-node/src/result_set.ts

```diff
@@ -35,33 +38,57 @@ export class ResultSet implements BaseResultSet<Stream.Readable> {
 
     validateStreamFormat(this.format)
 
-    let decodedChunk = ''
+    let incompleteChunks: Buffer[] = []
     const toRows = new Transform({
       transform(
         chunk: Buffer,
-        encoding: BufferEncoding,
+        _encoding: BufferEncoding,
         callback: TransformCallback
       ) {
-        decodedChunk += chunk.toString()
         const rows: Row[] = []
-        // eslint-disable-next-line no-constant-condition
-        while (true) {
-          const idx = decodedChunk.indexOf('\n')
-          if (idx !== -1) {
-            const text = decodedChunk.slice(0, idx)
-            decodedChunk = decodedChunk.slice(idx + 1)
-            rows.push({
-              text,
-              json<T>(): T {
-                return decode(text, 'JSON')
-              },
-            })
+        let lastIdx = 0
+        // first pass on the current chunk
+        // using the incomplete row from the previous chunks
+        let idx = chunk.indexOf(NEWLINE)
+        if (idx !== -1) {
```

Comment:
Do you think we could extract some of the logic into a separate piece? It might be a little bit hard to reason about

Reply:
I will check what's possible in one of the follow-ups.